Only warn in case of CHOLMOD major version mismatch

nalimilan opened the pull request:
The check introduced in #10362 is actually too strict,
as ABI break only happen with mismatches in the major
version. Warning on minor version differences means
distributions cannot update SuiteSparse without
rebuiling Julia (#12841).

Also fix the incorrect wording introuced in 25eb444:
this message corresponds to the case where the version is
different, not necessarily older.

CC: @andreasnoack

nalimilan committed:
Only warn in case of CHOLMOD major version mismatch

The check introduced in #10362 is actually too strict,
as ABI break only happen with mismatches in the major
version. Warning on minor version differences means
distributions cannot update SuiteSparse without
rebuiling Julia (#12841). We also require at least
libcholmod 2.1.1.

Call dlsym_e instead of dlsym, which raises an error instead
of returning C_NULL.

Also fix the incorrect wording introuced in 25eb444:
this message corresponds to the case where the version is
different, not necessarily older.

nalimilan commented on base/sparse/cholmod.jl:
```diff
@@ -87,22 +87,22 @@ function __init__()
             tmp = Array(Cint, 3)
             ccall((:cholmod_version, :libcholmod), Cint, (Ptr{Cint},), version_array)
```
BTW, I don't understand why this line is here. `version_array` is declared `const` above, and already set during compilation using the very same call. Since it's supposed to hold the version used when building Julia, resetting it on init doesn't make any sense to me. Should I remove it?

andreasnoack replied:
`version_array` ends up with the linked version of CHOLMOD if possible and `tmp` has the version at build time. Therefore, I think we could just have
```jl
tmp = copy(version_array)
```
instead of 
```jl
tmp = Array(Cint, 3)
ccall((:jl_cholmod_version, :libsuitesparse_wrapper), Cint, (Ptr{Cint},), tmp)
```
but I think there is a problem with `version` which should be updated at run time.

I think we need these definitions at compile time, but I'm not completely sure, so it might be that they can just be defined in the ``__init__`` function.

ivarne commented on base/sparse/cholmod.jl:
```diff
@@ -87,22 +87,22 @@ function __init__()
             ccall((:cholmod_version, :libcholmod), Cint, (Ptr{Cint},), version_array)
             ccall((:jl_cholmod_version, :libsuitesparse_wrapper), Cint, (Ptr{Cint},), tmp)
-            if tmp != version_array
+            if tmp[1] != version_array[1]
```
Doesn't this verify ony `x` (not `x` and `y`) in a `x.y.z` version number as @andreasnoack suggested in https://github.com/JuliaLang/julia/pull/10362#issuecomment-135881222. 

I don't know what policy COLMOD has for breaking changes, but I'd recommend verifying before loosening the requirement too much.

andreasnoack replied:
I'm not sure that the versioning policy in SuiteSaprse has been strict enough that testing only the major version is enough (though I initially thought that). E.g. the version number function for CHOLMOD was introduced in 2.1.0. I briefly looked at the changelog where I saw that
```
so although it is a major version number change for CHOLMOD it is not for the other libraries even though the API for providing custom memory allocation function has changed.


March 26, 2014: SuiteSparse version 4.3.0
    MAJOR UPDATE: first release of CHOLMOD GPU acceleration
    minor update: modified all packages to use SuiteSparse_config for
        malloc/calloc/realloc/free, printf, hypot, and divcomplex
        function pointers.

    * AMD 2.4.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * CAMD 2.4.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * CCOLAMD 2.9.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * CHOLMOD 3.0.0:
        MAJOR UPDATE to GPU acceleration.  Released at GTC 2014.
    * COLAMD 2.9.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * CSparse 3.1.3:
        minor update to UFget
    * CXSparse 3.1.3:
        minor update to UFget
    * KLU 1.3.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * LDL 2.2.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * RBio 2.2.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * SPQR 1.3.2:
        changed malloc/printf pointers to use SuiteSparse_config
    * UMFPACK 5.7.0:
        changed malloc/printf pointers to use SuiteSparse_config
    * MATLAB_Tools:
        added stats to UFcollection, revised commends in SFMULT,
        minor update to spok
```

nalimilan replied:
Ah, forgot to say I had checked the changelog before making this PR. I think we need to separate two things, which are basically what every sane library versioning policy follows:
1) Julia cannot use a different cholmod major version than the one it was built with because of ABI compatibility
2) Julia also has a minimum version requirement: it needs at least the first version in which all the used functions are present

1) is basically what the linker gives you automatically in C code (requiring the major SOVERSION). This is what this PR does. 2) is what is always handled separately by `configure` scripts, by package managers, or manually in Makefiles: just check that the version is above the minimum; most C programs don't check anything on their own though, and will just crash when they encounter an undefined symbol.

Of course we can do better than that. I can add another commit in which I check that the version is beyond a minimum. I can also have a look to find out what version that corresponds to.

andreasnoack replied:
> Of course we can do better than that. I can add another commit in which I check that the version is beyond a minimum. I can also have a look to find out what version that corresponds to.

That would be great.

nalimilan replied:
OK, I've made some tests, and I wasn't able to get the tests to pass with anything below SuiteSparse 4.2.0. With 4.1.0 I get this error (after rebuilding suitesparse-wrapper and the Julia image):
```
     * sparse               /home/milan/Dev/julia/usr/bin/julia: symbol lookup error: /home/milan/Dev/julia/usr/bin/../lib/libcholmod.so: undefined symbol: dpotrf_
```

This symbol is present in liblapack and can be resolved from Julia, so I'm not sure what's going on. If there's any chance this might work somewhere, I can try to get past it, else we can decide to stick to that. It includes libcholmod 2.1.1 (the second release with `cholmod_version()`), which means we would really be able to detect when the version is too old.

I've also discovered that the version check didn't even work below 4.1.0. With libcholmod < 2.1.0, i.e. when   `cholmod_version()` does not exist, `Libdl.dlsym` raised an error, while the code expected it to return `C_NULL`. That's a typical case of a rare code path that we supposedly care about but nobody tested. I've changed it to `try`/`catch`.

Does the logic in the new commit sound correct to you?

nalimilan commented:
(Moving from adding comments inline as they get hidden after a new commit is added.)

More precisions. I was actually able to get the tests to pass with 4.1.0 (it was related to a build issue). But with 4.0.0 they don't, and I get this error, which seems more serious:
```
     * sparse               ERROR: LoadError: LoadError("/home/milan/Dev/julia/usr/share/julia/test/sparse.jl",6,LoadError("/home/milan/Dev/julia/usr/share/julia/test/sparsedir/cholmod.jl",106,ErrorException("test failed: 1 == 0\n in expression: unsafe_load(chma.p).is_ll == 0")))
```

So we should rely on either 4.1.0 (libcholmod 2.1.0) or 4.2.0 (libcholmod 2.1.1) as a minimum. Since only the latter allows getting the version number, I'm inclined to require it. Both versions were released in the same month 2013. That's what this PR currently does.

I'll also write to Tim Davis to ensure we never get ABI breaks inside a minor release in the future.

Finally, I forgot to say that currently, even with this PR, sparse support doesn't work at all when the warning message is printed: since `StdErr` is not defined, it triggers an error, and the initialization phase does not complete. So we also need to fix #12841 separately.

tkelman commented on base/sparse/cholmod.jl:
```diff
@@ -65,46 +67,49 @@ const version = VersionNumber(version_array...)
+		        e.g. solving systems of equations with \\.
+
+		        It is recommended that you use Julia with a recent version
+		        of CHOLMOD, or download the OS X or generic Linux binaries
```
Could we just say "binaries from www.julialang.org" without mentioning the OS here? MSYS2 may eventually have a Julia package where this could also apply.

nalimilan replied:
Sure, will fix when the PR is ready.

nalimilan replied:
Fixed in the new version.

JeffBezanson commented:
@andreasnoack @ViralBShah can somebody comment on the current state of this?

nalimilan commented:
I've changed it to use `Libdl.dlsym_e`.

There's still something I don't understand in the logic of the code: why do we call `cholmod_version` at the top level, i.e. at build time? If we only care about the version used during the build, we could call `jl_cholmod_version` unconditionally there. At the moment all that is quite confusing, since the global `version_array` and `version` can get out of sync, since the former is used to store the current CHOLMOD version in `__init__`. Do we even need these globals? Wouldn't calling `jl_cholmod_version` in `__init__`be enough to find out the build time version of CHOLMOD?

andreasnoack commented:
There used to be function definitions the depended on the version and therefore the constants were convenient. However, these conditions have been removed. It is likely that we don't need these constants at compile time anymore so feel free to try to remove them.

nalimilan committed:
Simplify logic checking CHOLMOD version

Use jl_cholmod_version only once to get the version
used during the build, and do not modify this constant
in __init__(). Rename variables to make them more explicit.

Also only register gc tracked allocator if current version
(and not build version) is recent enough.

nalimilan commented:
@andreasnoack OK, I've added a commit simplifying the logic.

nalimilan commented:
With the last commit, the fields of the `C_Factor` struct are determined based on the CHOLMOD version the SuiteSparse wrapper was compiled with. This means that merely recompiling the Julia image without recompiling this wrapper won't make it take into account a change in the CHOLMOD version. I considered that's not an issue since an ABI break without rebuilding would generate problems with the wrapper anyway, right?

andreasnoack commented:
I think your reasoning here makes good sense. Package maintainers like you are also the main audience for these checks so if they make sense from a packaging perspective, we should be good. Users who downloads the generic builds or compile from source won't have these problems.

I don't think I have any further comment. Is the pr ready now?

nalimilan commented:
That's ready from my side. I've squashed the two first commits, as they are tightly related overlap a lot when moving the messages around.

ViralBShah commented:
This looks ok to me too. Can merge once CI completes.

mschauer commented:
Congrats, the last issue with milestone 0.4.0. Nice that it falls on number 13000.

nalimilan commented:
Cool, I hadn't noticed that detail! So I have the privilege of hitting the green button. Thanks everybody for the review!

ScottPJones commented:
So, does that mean that a 0.4 will be branched now, and master will be 0.5-dev?

tkelman commented on base/sparse/cholmod.jl:
```diff
@@ -54,57 +56,54 @@ end
+        if current_version < CHOLMOD_MIN_VERSION
+		    warn("""
+
+		        CHOLMOD version incompatibility
```
these should be spaces, not tabs

nalimilan replied:
Ah, I hate editors when they don't respect the existing style of a file. I've pushed a fix on master, should I also push it on release-0.4 to ease future backports?

tkelman replied:
just did d523ddea70ee9ad8c34fb4968d9b6140bdc12370
